LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer
+{
+    using LoRaWan.NetworkServer.BasicsStation;
+    using Microsoft.Extensions.Caching.Memory;
+    using Microsoft.Extensions.Logging;
+    using System;
+    using System.Collections.Generic;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    public sealed class ConcentratorDeduplication : IConcentratorDeduplication, IDisposable
+    {
+        internal MemoryCache Cache { get; private set; }
+
+        private readonly ILogger<IConcentratorDeduplication> logger;
+        private static readonly TimeSpan CacheEntryExpiration = TimeSpan.FromMinutes(1);
+
+        public ConcentratorDeduplication(ILogger<IConcentratorDeduplication> logger)
+        {
+            Cache = new MemoryCache(new MemoryCacheOptions());
+            this.logger = logger;
+        }
+
+        public bool ShouldDrop(UpstreamDataFrame updf, StationEui stationEui)
+        {
+            if (updf == null) throw new ArgumentNullException(nameof(updf));
+
+            var key = CreateCacheKey(updf);
+            StationEui previousStation;
+
+            lock (Cache)
```

Comment:
I was looking at AddOrGetExisting to replace this bit but that looks tricky. I am not a big fan of locking here, as this will be called very often. I was wondering maybe we could always add it and if the return is null get the value? (as memache is supposed to be thread safe)
@atifaziz @p-schuler

Reply:
Absolutely, if possible we should avoid locking. 

Note that `AddOrGetExisting` is part of `System.Runtime.Caching` namespace while we are using the recommended [according to the docs](https://docs.microsoft.com/en-us/aspnet/core/performance/caching/memory?view=aspnetcore-5.0#systemruntimecachingmemorycache] `Microsoft.Extensions.Caching.Memory.MemoryCache` 

For the `Microsoft.Extensions.Caching.Memory.MemoryCache` there is no mention in [the docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.memorycache?view=dotnet-plat-ext-5.0) of it being thread-safe. Looking at [the source](https://source.dot.net/#Microsoft.Extensions.Caching.Memory/MemoryCache.cs,99099576db12fef5) there is also no locking primitives used (other than some `Interlocked` usages for size counters) so not sure we can avoid locks if we stick with this MemoryCache. 

Checking quickly also the source for the [other MemoryCache](https://source.dot.net/#System.Runtime.Caching/System/Runtime/Caching/MemoryCache.cs,85751dcd91d94103) it's also not clear to me how it's supposed to be "thread-safe" (in quotes because I think the term is overloaded and different people mean different things with it). 

/cc @atifaziz because he had a similar comment, to keep the discussion in a single place.

Reply:
Unfortunately, It's thread-safe by virtue of using [`ConcurrentDictionary<,>`](https://github.com/dotnet/runtime/blob/d40f560efbf4f85ec6d59892a6c4bafa39f66d19/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L26) for the entries, which is probably why you don't see direct use of any locking primitives.

Reply:
Removed the locking with 227773b30e564a620e3a8868aa2efad23710b28a. If there are any resources, would like to read more about it this cache and thread-safety (not the one part of `System.Runtime`).

I noticed another occurrence of locking on this memory cache in `LoRaADRInMemoryStore` (possibly in other places too)

Reply:
To sum-up our discussion with @atifaziz : we agree that the desired semantics of a deduplication is "whoever is first to read or write wins". Given that, we do not need pessimistic locking and it was correctly (based on our current knowledge) removed. The question of whether the class is or not thread-safe does not affect the solution.